cc_ubuntu_advantage: add ua_config in auto-attach

aciba90 opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because all PRs are squash merged -->

```
cc_ubuntu_advantage: add ua_config in auto-attach

- Fix and document how to unset ua_config props.
- Redact ua_config values as they could contain sensitive auth info.
- Increase unit test coverage.
```

## Additional Context
<!-- If relevant -->
https://warthogs.atlassian.net/browse/SC-1257
https://github.com/canonical/cloud-init/pull/1583#pullrequestreview-1110047636

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

blackboxsw commented on cloudinit/config/cc_ubuntu_advantage.py:
```diff
@@ -29,8 +30,8 @@
         list is present, only named services will be activated. Whereas
-        'enable' list is not present, any named service will supplement
-        contract-default enabled services.
+        'enable' list is not present, the contract-default services will be
+        enabled.
 
```
```suggestion
        if the 'enable' list is not present, the contract's default
        services will be enabled.

```

aciba90 replied:
done

blackboxsw commented on cloudinit/config/cc_ubuntu_advantage.py:
```diff
@@ -172,102 +181,103 @@ def supplemental_schema_validation(ua_config: dict):
-                        f" Found: {value}"
-                    )
-            except (AttributeError, ValueError):
+        if key not in UA_CONFIG_PROPS:
+            LOG.warning(
+                "Ignoring unknown ubuntu_advantage.config.%s property", key
+            )
+            continue
```
While this is a nice sentiment, this now tightly couples `ubuntu_advantage: config` to whatever cloud-init releases. UA currently has more config options beyond the *proxy variables, such as update_status_timer, update_messaging_timer and metering_timer. I think we still want to allow setting those values, but maybe we just log  a warning and say, "Attempting to set unknown ua config option X".  If we are too strict here, customers which may want to immediately test ua config options that cloud-init doesn't know about will get blocked and have to wait for a cloud-init SRU to support such config settings.   And our worst-case scenario is that we can provide warnings for keys which may not be supported by the underlying ubuntu-advantage-tools version, and the ua config set <invalid_name> <value> will result in a traceback, which we want to fail cloud-init in error state if unsupported.

aciba90 replied:
done

blackboxsw commented on cloudinit/config/cc_ubuntu_advantage.py:
```diff
@@ -172,102 +181,103 @@ def supplemental_schema_validation(ua_config: dict):
-    # UA Config
-    for key, value in sorted(config.items()):
+    for key, value in sorted(ua_config.items()):
+        if key not in UA_CONFIG_PROPS:
+            continue  # Already logged in `supplemental_schema_validation`
```
I don't think we want to be so strict here (per my point above) there are some config settings already supported by UA that we might want to allow. This initial validation we were doing was to perform additional level of validation of known URL-type *proxy ua config values. If we want we can add the *_timer config options to cloud-init's schema, and assert that they are integers. But I still think we want to add an `additionalProperties: true`  under `ubuntu_advantage: config` in our schema-cloud-config-v1.json.

blackboxsw commented on cloudinit/config/schemas/schema-cloud-config-v1.json:
```diff
@@ -2229,37 +2229,38 @@
               "type": "object",
-              "description": "Configuration settings or override Ubuntu Advantage config. Only used in non Ubuntu Pro instances.",
+              "description": "Configuration settings or override Ubuntu Advantage config.",
+              "additionalProperties": false,
```
I actually think we want this to be true to allow for flexibility since UA currently has more config options than cloud-inti schema reports (since we only wanted to validate URL *proxy type keys). We also don't want to disallow valid confg keys as UA grows more because that would require cloud-init to continue to SRU to enable any new config keys that are supported.

aciba90 replied:
done

blackboxsw reviewed (changes requested):
Thanks @aciba90 . I hope we can allow cloud-init to be more premissive of additionalProperties, but I otherwise think the additional explicit logging about unknown config keys and handling is an improvement.

aciba90 commented:
> Thanks @aciba90 . I hope we can allow cloud-init to be more premissive of additionalProperties, but I otherwise think the additional explicit logging about unknown config keys and handling is an improvement.

Make sense. I have addressed the comments. Thanks, @blackboxsw, for the review.

blackboxsw reviewed (approved):
LGTM. The logging and handling is indeed much cleaner. Thanks for additionalProperties:true to allow us that flexibility to decouple UA releases from cloud-init SRUs.
